packages/flutter/lib/src/widgets/widget_state.dart

```diff
@@ -0,0 +1,653 @@
+///    `WidgetStateProperty<Color>`.
+///  * [WidgetStateMouseCursor], a [MouseCursor] that implements
+///    `WidgetStateProperty` which is used in APIs that need to accept either
+///    a [MouseCursor] or a [WidgetStateProperty<MouseCursor>].
+///  * [WidgetStateOutlinedBorder], an [OutlinedBorder] that implements
+///    `WidgetStateProperty` which is used in APIs that need to accept either
+///    an [OutlinedBorder] or a [WidgetStateProperty<OutlinedBorder>].
+///  * [WidgetStateBorderSide], a [BorderSide] that implements
+///    `WidgetStateProperty` which is used in APIs that need to accept either
+///    a [BorderSide] or a [WidgetStateProperty<BorderSide>].
+///  * [WidgetStateTextStyle], a [TextStyle] that implements
+///    `WidgetStateProperty` which is used in APIs that need to accept either
+///    a [TextStyle] or a [WidgetStateProperty<TextStyle>].
+/// {@endtemplate}
+enum WidgetState {
+  /// The state when the user drags their mouse cursor over the given widget.
+  ///
+  /// See: https://material.io/design/interaction/states.html#hover.
+  hovered,
+
+  /// The state when the user navigates with the keyboard to a given widget.
+  ///
+  /// This can also sometimes be triggered when a widget is tapped. For example,
+  /// when a [TextField] is tapped, it becomes [focused].
+  ///
+  /// See: https://material.io/design/interaction/states.html#focus.
+  focused,
+
+  /// The state when the user is actively pressing down on the given widget.
+  ///
+  /// See: https://material.io/design/interaction/states.html#pressed.
+  pressed,
+
+  /// The state when this widget is being dragged from one place to another by
+  /// the user.
+  ///
+  /// https://material.io/design/interaction/states.html#dragged.
+  dragged,
+
+  /// The state when this item has been selected.
+  ///
+  /// This applies to things that can be toggled (such as chips and checkboxes)
+  /// and things that are selected from a set of options (such as tabs and radio buttons).
+  ///
+  /// See: https://material.io/design/interaction/states.html#selected.
+  selected,
+
+  /// The state when this widget overlaps the content of a scrollable below.
+  ///
+  /// Used by [AppBar] to indicate that the primary scrollable's
+  /// content has scrolled up and behind the app bar.
+  scrolledUnder,
+
+  /// The state when this widget is disabled and cannot be interacted with.
+  ///
+  /// Disabled widgets should not respond to hover, focus, press, or drag
+  /// interactions.
+  ///
+  /// See: https://material.io/design/interaction/states.html#disabled.
+  disabled,
+
+  /// The state when the widget has entered some form of invalid state.
+  ///
+  /// See https://material.io/design/interaction/states.html#usage.
+  error,
+}
+
+/// Signature for the function that returns a value of type `T` based on a given
+/// set of states.
+typedef WidgetPropertyResolver<T> = T Function(Set<WidgetState> states);
+
+/// Defines a [Color] that is also a [WidgetStateProperty].
```

Comment:
This API doc, which I assume is roughly a template for similar classes, is great.